torchgeo/datasets/geo.py

```diff
@@ -676,46 +780,70 @@ def _load_image(self, index: int) -> Tuple[Tensor, Tensor]:
         return tensor, label
 
 
-class ZipDataset(GeoDataset):
-    """Dataset for merging two or more GeoDatasets.
+class IntersectionDataset(GeoDataset):
+    """Dataset representing the intersection of two GeoDatasets.
+
+    This allows users to do things like:
+
+    * Combine image and target labels and sample from both simultaneously
+      (e.g. Landsat and CDL)
+    * Combine datasets for multiple image sources for multimodal learning or data fusion
+      (e.g. Landsat and Sentinel)
+
+    These combinations require that all queries are present in *both* datasets,
+    and can be combined using an :class:`IntersectionDataset`:
 
-    For example, this allows you to combine an image source like Landsat8 with a target
-    label like CDL.
+    .. code-block:: python
+
+       dataset = landsat & cdl
     """
 
-    def __init__(self, datasets: Sequence[GeoDataset]) -> None:
```

Comment:
Previously this took a list of datasets as input. I switched it to take exactly 2 datasets as input as it makes computing the intersection of the indices much easier. Personally I don't think users should use `IntersectionDataset` or `UnionDataset` directly, they should use the `&` and `|` operators. Since these operators are binary, we only really need to support exactly 2 datasets at a time.